File: 2025/NeetCode/Arrays & Hashing/49. Group Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
 
    let map = {};

    for(let i of strs) {
        let key = i.split('').sort().join('');

        if(!map[key]) {
            map[key] = [i];
        } else {
            map[key].push(i);
        }
    } 

    return(Object.values(map));
    
};

// Time Complexity = O(m * nlogn) where m = number of words in strs | n = number of characters in each word
// Space Complexity = O(n)

File: 2026/Blind 75 [Python]/Arrays & Hashing/49. Group Anagrams.py
from typing import List
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram_map = defaultdict(list)
        result = []

        for s in strs:
            sorted_s = tuple(sorted(s))
            anagram_map[sorted_s].append(s)

        for value in anagram_map.values():
            result.append(value)

        return result

# Time Complexity: O(N * KlogK)
# Space Complexity: O(N * K)

# N = number of strings
# K = max length of each string